pallets/omnipool/src/lib.rs

```diff
@@ -0,0 +1,1337 @@
+			)?;
+
+			let delta_imbalance = Self::recalculate_imbalance(&new_asset_state, state_changes.delta_imbalance)
+				.ok_or(ArithmeticError::Overflow)?;
+			Self::update_imbalance(<HubAssetImbalance<T>>::get(), delta_imbalance)?;
+
+			Self::update_tvl(&state_changes.asset.delta_tvl)?;
+
+			Self::update_hub_asset_liquidity(&state_changes.asset.delta_hub_reserve)?;
+
+			Self::set_asset_state(asset, new_asset_state);
+
+			Self::deposit_event(Event::LiquidityAdded {
+				from: who,
+				asset_id: asset,
+				amount,
+				position_id: instance_id,
+			});
+			Ok(())
+		}
+
+		/// Remove liquidity of asset `asset` in quantity `amount` from Omnipool
+		///
+		/// `remove_liquidity` removes specified shares amount from given PositionId (NFT instance).
+		///
+		/// if all shares from given position are removed, NFT is burned.
+		///
+		/// Parameters:
+		/// - `position_id`: The identifier of position which liquidity is removed from.
+		/// - `amount`: Amount of shares removed from omnipool
+		///
+		/// Emits `LiquidityRemoved` event when successful.
+		///
+		#[pallet::weight(<T as Config>::WeightInfo::remove_liquidity())]
+		#[transactional]
+		pub fn remove_liquidity(
+			origin: OriginFor<T>,
+			position_id: T::PositionInstanceId,
+			amount: Balance,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(
+				T::NFTHandler::owner(&T::NFTClassId::get(), &position_id) == Some(who.clone()),
+				Error::<T>::Forbidden
+			);
+
+			let position = Positions::<T>::get(position_id).ok_or(Error::<T>::PositionNotFound)?;
+
+			ensure!(position.shares >= amount, Error::<T>::InsufficientShares);
+
+			let stable_asset = Self::stable_asset()?;
+
+			let asset_id = position.asset_id;
+
+			let asset_state = Self::load_asset_state(asset_id)?;
+
+			let state_changes = calculate_remove_liquidity_state_changes::<T::AssetId>(
+				&asset_state,
+				amount,
+				&position,
+				stable_asset,
+				asset_id == T::StableCoinAssetId::get(),
+			)
+			.ok_or(ArithmeticError::Overflow)?;
+
+			let new_asset_state = asset_state
+				.delta_update(&state_changes.asset)
+				.ok_or(ArithmeticError::Overflow)?;
+
+			// Update position state
+			let updated_position = position
+				.delta_update(
+					&state_changes.delta_position_reserve,
+					&state_changes.delta_position_shares,
+				)
+				.ok_or(ArithmeticError::Overflow)?;
+
+			// Token balance updates
+			T::Currency::transfer(
+				asset_id,
+				&Self::protocol_account(),
+				&who,
+				*state_changes.asset.delta_reserve,
+			)?;
+
+			let delta_imbalance = Self::recalculate_imbalance(&new_asset_state, state_changes.delta_imbalance)
+				.ok_or(ArithmeticError::Overflow)?;
+			Self::update_imbalance(<HubAssetImbalance<T>>::get(), delta_imbalance)?;
+
+			Self::update_tvl(&state_changes.asset.delta_tvl)?;
+
+			Self::update_hub_asset_liquidity(&state_changes.asset.delta_hub_reserve)?;
+
+			// LP receives some hub asset
+			if state_changes.lp_hub_amount > Balance::zero() {
```

Comment:
there are definitely tests where this is 0. but again it might not be obvious as there is not dedicated test for this. 

but it wll be when tests are refactored.